orm/encoding/ormvalue/codec.go

```diff
@@ -0,0 +1,103 @@
+package ormvalue
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/cosmos/cosmos-sdk/orm/types/ormerrors"
+
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// Codec defines an interface for decoding and encoding values in ORM index keys.
+type Codec interface {
+
+	// Decode decodes a value in a key.
+	Decode(r *bytes.Reader) (protoreflect.Value, error)
+
+	// Encode encodes a value in a key.
+	Encode(value protoreflect.Value, w io.Writer) error
+
+	// Compare compares two values of this type and should primarily be used
+	// for testing.
+	Compare(v1, v2 protoreflect.Value) int
+
+	// IsOrdered returns true if callers can always assume that this ordering
+	// is suitable for sorted iteration.
+	IsOrdered() bool
+
+	// FixedSize returns a positive value if encoders should assume a fixed size
+	// buffer for encoding. A codec may choose to estimate this value,
+	// in the case of a varint for example, so that computation with Size isn't
+	// needed. Encoders should only use the bytes actually written
+	// by Encode.
+	FixedSize() int
+
+	// Size estimates the size needed to encode the field. It may over-estimate
+	// by a small amount if this is more efficient, for instance in the case of
+	// a varint. Encoders should only use the bytes actually written
+	// by Encode.
+	Size(value protoreflect.Value) (int, error)
+}
+
+var (
+	timestampMsgType  = (&timestamppb.Timestamp{}).ProtoReflect().Type()
+	timestampFullName = timestampMsgType.Descriptor().FullName()
+	durationMsgType   = (&durationpb.Duration{}).ProtoReflect().Type()
+	durationFullName  = durationMsgType.Descriptor().FullName()
+)
+
+// GetCodec returns the Codec for the provided field if one is defined.
+// nonTerminal should be set to true if this value is being encoded as a
+// non-terminal segment of a multi-part key.
+func GetCodec(field protoreflect.FieldDescriptor, nonTerminal bool) (Codec, error) {
+	if field.IsList() {
```

Comment:
Missing case for map?

Reply:
Map should fail under KIND_MESSAGE

Reply:
Either way we should probably have tests for failure cases